packages/form-core/src/FormApi.ts

```diff
@@ -393,103 +387,172 @@ export class FormApi<
+        const prevVal = _prevVal as FormState<TFormData> | undefined
+        const prevBaseStore = prevDepVals?.[0]
+        const currBaseStore = currDepVals[0]
+
+        // Computed state
+        const fieldMetaValues = Object.values(currBaseStore.fieldMetaBase) as (
+          | FieldMeta
+          | undefined
+        )[]
+
+        const isFieldsValidating = fieldMetaValues.some(
+          (field) => field?.isValidating,
+        )
 
-          const isTouched = fieldMetaValues.some((field) => field?.isTouched)
-          const isBlurred = fieldMetaValues.some((field) => field?.isBlurred)
+        const isFieldsValid = !fieldMetaValues.some(
+          (field) =>
+            field?.errorMap &&
+            isNonEmptyArray(Object.values(field.errorMap).filter(Boolean)),
+        )
+
+        const isTouched = fieldMetaValues.some((field) => field?.isTouched)
+        const isBlurred = fieldMetaValues.some((field) => field?.isBlurred)
 
+        const shouldInvalidateOnMount =
           // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-          if (isTouched && state?.errorMap?.onMount) {
-            state.errorMap.onMount = undefined
-          }
+          isTouched && currBaseStore?.errorMap?.onMount
 
-          const isDirty = fieldMetaValues.some((field) => field?.isDirty)
-          const isPristine = !isDirty
+        const isDirty = fieldMetaValues.some((field) => field?.isDirty)
+        const isPristine = !isDirty
 
-          const hasOnMountError = Boolean(
+        const hasOnMountError = Boolean(
+          // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+          currBaseStore.errorMap?.onMount ||
             // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-            state.errorMap?.onMount ||
-              // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-              fieldMetaValues.some((f) => f?.errorMap?.onMount),
-          )
+            fieldMetaValues.some((f) => f?.errorMap?.onMount),
+        )
 
-          const isValidating = isFieldsValidating || state.isFormValidating
-          state.errors = Object.values(state.errorMap).reduce((prev, curr) => {
-            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-            if (curr === undefined) return prev
-            if (typeof curr === 'string') {
-              prev.push(curr)
-              return prev
-            } else if (curr && isFormValidationError(curr)) {
-              prev.push(curr.form)
+        const isValidating = !!isFieldsValidating
+
+        // As `errors` is not a primitive, we need to aggressively persist the same referencial value for performance reasons
+        let errors = prevVal?.errors ?? []
+        if (
+          !prevBaseStore ||
+          currBaseStore.errorMap !== prevBaseStore.errorMap
+        ) {
+          errors = Object.values(currBaseStore.errorMap).reduce(
```

Comment:
Can we check earlier if `onMount` errors have to be removed? So that we avoid pushing them now and then filtering them out on lines 515-521

Something like moving up this block
```ts
let errorMap = currBaseStore.errorMap
if (shouldInvalidateOnMount) {
    errorMap = Object.assign(errorMap, { onMount: undefined })
}
```
and then using the local `errorMap` on `Object.values`

Reply:
Let's make a follow-up PR on this, seeing that it works today and I'm struggling to think of an easy way to move this around ATM

Reply:
Sure!